Sources/Hyperspace/Service/Backend/BackendService.swift

```diff
@@ -32,11 +34,29 @@ extension BackendService: BackendServiceProtocol {
     public func execute<T: NetworkRequest>(request: T, completion: @escaping BackendServiceCompletion<T.ResponseType, T.ErrorType>) {
         networkService.execute(request: request.urlRequest) { result in
             switch result {
-            case .success(let result):
-                BackendServiceHelper.handleResponseData(result.data, for: request, completion: completion)
-            case .failure(let result):
-                DispatchQueue.main.async {
-                    completion(.failure(T.ErrorType(networkServiceFailure: result)))
+            case .success(let serviceSuccess):
+                BackendServiceHelper.handleResponseData(serviceSuccess.data, for: request, completion: completion)
+            case .failure(let serviceFailure):
+                BackendServiceHelper.handleNetworkServiceFailure(serviceFailure, completion: completion)
+            }
+        }
+    }
+    
+    public func execute<T: NetworkRequest & Recoverable>(recoverable request: T, completion: @escaping BackendServiceCompletion<T.ResponseType, T.ErrorType>) {
+        execute(request: request) { [weak self] result in
+            switch result {
+            case .success(let response):
+                BackendServiceHelper.handleResponse(response, completion: completion)
+                
+            case .failure(let error):
+                guard let recoveryStrategy = self?.recoveryStrategy else { return completion(.failure(error)) }
```

Comment:
For the `else` on this `guard`, do we need to pipe it through `BackendServerHelper` so that the completion block gets run on the main thread?

Reply:
Good catch. We very much do. Will tweak that now.